# ハンバーガー統計学 with Python 第7章

第５回講義（2017年10月13日講義、10月17日課題提出締切）

このページは、<a href="http://kogolab.chillout.jp/elearn/hamburger/index.html" target="_blank">ハンバーガー統計学にようこそ！</a>の内容をプログラミング言語 Python に合わせて書き直したものです（著作権者の許諾を得て利用しております）。

<h1 STYLE="background: #c2edff;padding: 0.5em;"><a href="http://kogolab.chillout.jp/elearn/hamburger/chap7/sec0.html" target="_blank">第７章 分散分析（２要因）</a></h1>

ワクワク、モグモグ、そしてパクパクと、あなたの住んでいる町は、ハンバーガーショップの激戦地になりました。

その結果、ワクワクバーガーは 「打倒パクパク！」に燃えています。

そうこうしていたある日、ワクワクバーガーに行くと、新しいチキンを売り出したい、という話が持ち上がっていました。

「新メニューを売り出して、パクパクにがつーんと差をつけたいんだ！」

ということでワクワクが開発したのは、「クリスピーチキン」と「辛口チキン」でした。

<table><tr><td><img src="image/chicken00001.jpg"></td>
<td><img src="image/chicken00002.jpg"></td></tr></table>

<center>クリスピーチキン　　と　　辛口チキン</center>

しかし、ここまで話して店員さんが悩みを打ち明けました。

「実は、悩んでいるだけどね……」

<img src="image/salesman00022.jpg">

「というのは、クリスピーチキンと、辛口チキンを組み合わせるっていう案もでているんだよ。クリスピー辛口チキンのように組み合わせたほうがいいのかなぁ？　それともクリスピーと辛口は、それぞれ別々の商品にしたほうがいいのかなぁ？」

<table><tr><td><img src="image/chicken00001.jpg"></td>
<td>+</td>
<td><img src="image/chicken00002.jpg"></td>
<td>=</td>
<td><img src="image/chicken00003.jpg"></td>
</tr></table>

それとも、クリスピー辛口チキンがいいのか?

そこで、あなたは調べることにしました。



<h2 STYLE="background: #c2edff;padding: 0.5em;"><a href="http://kogolab.chillout.jp/elearn/hamburger/chap7/sec1.html" target="_blank">7.1　2つの要因では?</a></h2>

<h4 style="border-bottom: solid 1px black;">要因と水準</h4>

観測データに影響を与えそうな原因を「要因」と呼びます。この場合、クリスピーか普通の衣かという「食感」は、お客さんの好みに影響を与えそうなので、要因としてとらえられます。

また、要因の中の条件の違いのことを「水準」と呼びます。この場合は、食感がクリスピーであるか、普通の衣であるかということです。

表にしてみると次のようになります。

<table width="50%" border="1" align="center" cellpadding="0" cellspacing="0">
  <tr bgcolor="#CCCCCC"> 
    <td width="40%" bgcolor="#99CCFF"> 
      <div align="center">要因A</div></td>
    <td width="30%"> 
      <div align="center">水準a1</div></td>
    <td> 
      <div align="center">水準a2</div></td>
  </tr>
  <tr> 
    <td><div align="center">食感</div></td>
    <td><div align="center">クリスピー</div></td>
    <td><div align="center">普通の衣</div></td>
  </tr>
</table>

また、同じように考えると、辛口か普通味かという「味付け」も要因としてとらえられます。

この要因の水準は、辛口か普通味かということになります。次の表のようになります。

<table width="50%" border="1" align="center" cellpadding="0" cellspacing="0">
  <tr bgcolor="#CCCCCC"> 
    <td width="40%" bgcolor="#FFCC99"> 
      <div align="center">要因B</div></td>
    <td width="30%"> <div align="center">水準b1</div></td>
    <td> <div align="center">水準b2</div></td>
  </tr>
  <tr> 
    <td><div align="center">味付け</div></td>
    <td><div align="center">辛口</div></td>
    <td><div align="center">普通味</div></td>
  </tr>
</table>

つまり、要因が2つあるということです。それぞれの要因について、2つの水準がありますので、次の表のように、全部で4種類のチキンが考えられます。

<table border="1" align="center" cellpadding="0" cellspacing="0">
  <tr> 
    <td bgcolor="#99CCFF"> 
      <div align="center">要因A（食感）</div></td>
    <td colspan="2" bgcolor="#CCCCCC">
<div align="center">クリスピー（水準a1）</div></td>
    <td colspan="2" bgcolor="#CCCCCC">
<div align="center">普通の衣（水準a2）</div></td>
  </tr>
  <tr> 
    <td bgcolor="#FFCC99">
<div align="center">要因B（味付け）</div></td>
    <td bgcolor="#CCCCCC">
<div align="center">辛口<br>（水準b1）</div></td>
    <td bgcolor="#CCCCCC">
<div align="center">普通味<br>（水準b2）</div></td>
    <td width="20%" bgcolor="#CCCCCC">
<div align="center">辛口<br>（水準b1）</div></td>
    <td width="20%" bgcolor="#CCCCCC">
<div align="center">普通味<br>（水準b2）</div></td>
  </tr>
  <tr> 
    <td width="20%"><div align="center">群（種類）</div></td>
    <td width="20%"><div align="center">クリスピーで辛口 (a1b1群)</div></td>
    <td width="20%"><div align="center">クリスピーで普通味 (a1b2群)</div></td>
    <td><div align="center">普通の衣で辛口 (a2b1群)</div></td>
    <td><div align="center">普通の衣で普通味 (a2b2群)</div></td>
  </tr>
</table>

ところで、前の6章では、要因はお店の違いでした。そして、水準は、ワクワク、モグモグ、パクパクの3つがありました。こんな感じです。

<table border="1" align="center" cellpadding="0" cellspacing="0">
  <tr bgcolor="#CCCCCC"> 
    <td width="25%" bgcolor="#99CCFF"> 
      <div align="center">要因</div></td>
    <td width="25%"> 
      <div align="center">水準1</div></td>
    <td width="25%"> 
      <div align="center">水準2</div></td>
    <td><div align="center">水準3</div></td>
  </tr>
  <tr> 
    <td><div align="center">お店の違い</div></td>
    <td><div align="center">ワクワク</div></td>
    <td><div align="center">モグモグ</div></td>
    <td><div align="center">パクパク</div></td>
  </tr>
</table>

これを「１要因の分散分析」と呼びます。要因が１つだからです。

今回の分散分析は、「２要因の分散分析」です。要因が２つだからです。

<h4 style="border-bottom: solid 1px black;">データを集める</h4>

さて、店員さんに4種類のチキンをそれぞれ15個ずつ作ってもらい、街の人60人に食べてもらい、そのおいしさについて100点満点で点数をつけてもらいました。

そのデータは次のようになりました。

<TABLE WIDTH="451" HEIGHT="345" BORDER="1" CELLSPACING="0" 
  CELLPADDING="0">
    <TR> 
      <TD height="23" COLSPAN="2" BGCOLOR="#eeee77"> 
        <CENTER>
          クリスピー </CENTER></TD>
      <TD COLSPAN="2" BGCOLOR="#CCCCCC"> 
        <CENTER>
          普通の衣 
        </CENTER></TD>
    </TR>
    <TR> 
      <TD WIDTH="20%" BGCOLOR="#ff8888" HEIGHT="20"> <CENTER>
          辛口 </CENTER></TD>
      <TD WIDTH="20%" BGCOLOR="#dddddd"> <CENTER>
          普通味 
        </CENTER></TD>
      <TD WIDTH="20%" BGCOLOR="#ff8888"> <CENTER>
          辛口 </CENTER></TD>
      <TD WIDTH="20%" BGCOLOR="#dddddd"> <CENTER>
          普通味 
        </CENTER></TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER" HEIGHT="20"> 65</TD>
      <TD ALIGN="CENTER"> 65</TD>
      <TD ALIGN="CENTER"> 70</TD>
      <TD ALIGN="CENTER"> 70</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER" HEIGHT="20"> 85</TD>
      <TD ALIGN="CENTER"> 70</TD>
      <TD ALIGN="CENTER"> 65</TD>
      <TD ALIGN="CENTER"??> 70</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER" HEIGHT="20"> 75</TD>
      <TD ALIGN="CENTER"> 80</TD>
      <TD ALIGN="CENTER"> 85</TD>
      <TD ALIGN="CENTER"> 85</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER" HEIGHT="20"> 85</TD>
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER"> 80</TD>
      <TD ALIGN="CENTER"> 80</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER" HEIGHT="20"> 75</TD>
      <TD ALIGN="CENTER"> 70</TD>
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER"> 65</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER" HEIGHT="20"> 80</TD>
      <TD ALIGN="CENTER"> 60</TD>
      <TD ALIGN="CENTER"> 65</TD>
      <TD ALIGN="CENTER"> 75</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER" HEIGHT="20"> 90</TD>
      <TD ALIGN="CENTER"> 65</TD>
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER"> 65</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER" HEIGHT="20"> 75</TD>
      <TD ALIGN="CENTER"> 70</TD>
      <TD ALIGN="CENTER"> 60</TD>
      <TD ALIGN="CENTER"> 85</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER" HEIGHT="20"> 85</TD>
      <TD ALIGN="CENTER"> 85</TD>
      <TD ALIGN="CENTER"> 85</TD>
      <TD ALIGN="CENTER"> 80</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER" HEIGHT="20"> 65</TD>
      <TD ALIGN="CENTER"> 60</TD>
      <TD ALIGN="CENTER"> 65</TD>
      <TD ALIGN="CENTER"> 60</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER" HEIGHT="20"> 75</TD>
      <TD ALIGN="CENTER"> 65</TD>
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER"> 70</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER" HEIGHT="20"> 85</TD>
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER"> 70</TD>
      <TD ALIGN="CENTER"> 75</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER" HEIGHT="20"> 80</TD>
      <TD ALIGN="CENTER"> 70</TD>
      <TD ALIGN="CENTER"> 65</TD>
      <TD ALIGN="CENTER"> 70</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER" HEIGHT="20"> 85</TD>
      <TD ALIGN="CENTER"> 80</TD>
      <TD ALIGN="CENTER"> 80</TD>
      <TD ALIGN="CENTER"> 80</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER" HEIGHT="20"> 90</TD>
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER"> 85</TD>
    </TR>
  </TABLE>

In [1]:
# 以下の４つをそれぞれ「群」と呼びます。
a1b1 = [65, 85, 75, 85, 75, 80, 90, 75, 85, 65, 75, 85, 80, 85, 90] # クリスピーで辛口
a1b2 = [65, 70, 80, 75, 70, 60, 65, 70, 85, 60, 65, 75, 70, 80, 75] # クリスピーで普通味
a2b1 = [70, 65, 85, 80, 75, 65, 75, 60, 85, 65, 75, 70, 65, 80, 75] # 普通の衣で辛口
a2b2 = [70, 70, 85, 80, 65, 75, 65, 85, 80, 60, 70, 75, 70, 80, 85] # 普通の衣で普通味

それぞれの平均と標準偏差を計算してみましょう。

In [2]:
# 平均値を求める関数
def average(list):
    sum = 0
    for x in list:
        sum += x
    return sum / len(list)

In [3]:
# 分散を計算する
def variance(list):
    ave = average(list)
    sum = 0
    for x in list:
        sum += (x - ave) ** 2
    return sum / len(list)

In [4]:
# 標準偏差を計算する
import math
def standard_deviation(list):
    return math.sqrt(variance(list))

In [5]:
print(average(a1b1), standard_deviation(a1b1))
print(average(a1b2), standard_deviation(a1b2))
print(average(a2b1), standard_deviation(a2b1))
print(average(a2b2), standard_deviation(a2b2))

79.66666666666667 7.630348761506397
71.0 7.118052168020874
72.66666666666667 7.4981479194679945
74.33333333333333 7.71722460186015


<h4 style="border-bottom: solid 1px black;">ズレの分解</h4>

前回の１要因の分散分析の場合は、ズレを次のように分解しました。

<img src="image/anova4.jpg">

このように考えて、全体の平均からのズレを、群間のズレと群内のズレに分解しました。

全体の平均からのズレ　＝　群間のズレ　＋　群内のズレ

ということです。

さて、今回は、要因が２つあるので、ちょっと複雑になります。

要因が２つありますので、食感の要因（クリスピーか普通の衣か）によるズレと、味付けの要因（辛口か普通味か）によるズレが考えられます。

全体の平均からのズレ　＝　食感の要因によるズレ　＋　味付けの要因によるズレ　＋　残りのズレ（残差）

となります。残りのズレのことを「残差」と呼びましょう。これは１要因の分散分析では「群内の残差」に当たります。

また、１つの要因の単独の効果を「主効果」といいます。ここでは、食感の要因による主効果と、味付けの要因による主効果の２つの主効果があります。

<h4 style="border-bottom: solid 1px black;">交互作用</h4>

残差はこの３つだけで良いでしょうか？

いえ、不足しています。たまたま、食感（クリスピーか普通の衣）にかかわらず、味付けの要因による残差が一定ならば、この式であっています。しかし、そうでないことも多いのです。

つまり、「食感の要因と味付けの要因の２つが組み合わさって生じる残差」を考える必要があります。ちょっとわかりにくいですが、こう考えてください。

* 「衣がクリスピーだろうが、普通だろうが、味付けを辛口にすると一定の効果がある」というのであれば、主効果による残差だけでよい。
* しかし、そうでないことも多い。
* たとえば、衣がクリスピーのときに味付けを辛口にすると、効果が大きいけれども、普通の衣の時に味付けを辛口にしても、それほど効果は大きくない、というような場合だ。
* つまり、食感の要因と味付けの要因の組み合わせによって効果が変わることも多い。
* したがって、要因の組み合わせによる残差を考えに入れる必要がある。

組み合わせによる残差を式にいれてみると、次のようになります。

全体の平均からの残差　＝　食感の要因による残差　＋　味付けの要因による残差　＋　食感の要因と味付けの要因の組み合わせによる残差　＋　残りの残差

２つの要因の組み合わせによって起こる効果を「交互作用」といいます。

もっと簡単に書き直すと次のようになります。

全体の残差　＝　要因Aによる残差　＋　要因Bによる残差　＋　交互作用による残差　＋　残りの残差

交互作用については、重要ですので、あとでもう一度説明します。

<h4 style="border-bottom: solid 1px black;">２要因の分散分析</h4>

２要因の分散分析の考え方を説明します。

全体の残差　＝　要因Aによる残差　＋　要因Bによる残差　＋　交互作用による残差　＋　残りの残差

ということなので、残りの残差を規準にして、要因Aによる残差、要因Bによる残差、交互作用による残差の３つの大きさを検討します。

* 残りの残差に対して、要因Aによる残差が大きければ、主効果Aが大きいことが分かります。
* 残りの残差に対して、要因Bによる残差が大きければ、主効果Bが大きいことが分かります。
* 残りの残差に対して、交互作用による残差が大きければ、交互作用が大きいことが分かります。

２要因の分散分析の帰無仮説は、

「要因Aによる差がなく、要因Bによる差がなく、また交互作用による差もない」

となります。

そうすると対立仮説はその否定で、

「要因Aによる差があるか、要因Bによる差があるか、または、交互作用による差があるか、どれか１つが成り立つ」

となります。

これは、具体的には、次のような7つの場合があります。

* Aの主効果だけが有意
* Bの主効果だけが有意
* 交互作用だけが有意
* Aの主効果とBの主効果が有意
* Aの主効果と交互作用が有意
* Bの主効果と交互作用が有意
* Aの主効果とBの主効果と交互作用のすべてが有意

これを検定するのが、２要因の分散分析ということになります。

<h2 STYLE="background: #c2edff;padding: 0.5em;"><a href="http://kogolab.chillout.jp/elearn/hamburger/chap7/sec2.html" target="_blank">7.2　２要因の分散分析表</a></h2>

それでは具体的に計算していきましょう。

まず、１要因の分散分析でやったように、各群のデータ数、平均、標準偏差、分散、平方和を計算します。

 <TABLE HEIGHT="345" BORDER="1" CELLSPACING="0" 
  CELLPADDING="0">
    <TR> 
      <td bgcolor="#cccccc"></td>
      <TD BGCOLOR="#dddddd" HEIGHT="20"> <CENTER>
          クリスピーで辛口 </CENTER></TD>
      <TD BGCOLOR="#dddddd"> <CENTER>
          クリスピーで普通味 
        </CENTER></TD>
      <TD BGCOLOR="#dddddd"> <CENTER>
          普通の衣で辛口 </CENTER></TD>
      <TD WIDTH="20%" BGCOLOR="#dddddd"> <CENTER>
          普通の衣で普通味 
        </CENTER></TD>
      <TD BGCOLOR="#dddddd" HEIGHT="20"> <CENTER>
          クリスピー </CENTER></TD>
      <TD BGCOLOR="#dddddd"> <CENTER>
          普通の衣 
        </CENTER></TD>
      <TD BGCOLOR="#dddddd"> <CENTER>
          辛口 </CENTER></TD>
      <TD BGCOLOR="#dddddd"> <CENTER>
          普通味 
        </CENTER></TD>
      <TD BGCOLOR="#dddddd"> <CENTER>
          全体 
        </CENTER></TD>
    </TR>
    <TR> 
      <td bgcolor="#cccccc">1</td>
      <TD ALIGN="CENTER" HEIGHT="20"> 65</TD>
      <TD ALIGN="CENTER"> 65</TD>
      <TD ALIGN="CENTER"> 70</TD>
      <TD ALIGN="CENTER"> 70</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR> 
      <td bgcolor="#cccccc">2</td>
      <TD ALIGN="CENTER" HEIGHT="20"> 85</TD>
      <TD ALIGN="CENTER"> 70</TD>
      <TD ALIGN="CENTER"> 65</TD>
      <TD ALIGN="CENTER"??> 70</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR> 
      <td bgcolor="#cccccc">3</td>
      <TD ALIGN="CENTER" HEIGHT="20"> 75</TD>
      <TD ALIGN="CENTER"> 80</TD>
      <TD ALIGN="CENTER"> 85</TD>
      <TD ALIGN="CENTER"> 85</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR> 
      <td bgcolor="#cccccc">4</td>
      <TD ALIGN="CENTER" HEIGHT="20"> 85</TD>
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER"> 80</TD>
      <TD ALIGN="CENTER"> 80</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR> 
      <td bgcolor="#cccccc">5</td>
      <TD ALIGN="CENTER" HEIGHT="20"> 75</TD>
      <TD ALIGN="CENTER"> 70</TD>
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER"> 65</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR> 
      <td bgcolor="#cccccc">6</td>
      <TD ALIGN="CENTER" HEIGHT="20"> 80</TD>
      <TD ALIGN="CENTER"> 60</TD>
      <TD ALIGN="CENTER"> 65</TD>
      <TD ALIGN="CENTER"> 75</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR> 
      <td bgcolor="#cccccc">7</td>
      <TD ALIGN="CENTER" HEIGHT="20"> 90</TD>
      <TD ALIGN="CENTER"> 65</TD>
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER"> 65</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR> 
      <td bgcolor="#cccccc">8</td>
      <TD ALIGN="CENTER" HEIGHT="20"> 75</TD>
      <TD ALIGN="CENTER"> 70</TD>
      <TD ALIGN="CENTER"> 60</TD>
      <TD ALIGN="CENTER"> 85</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR> 
      <td bgcolor="#cccccc">9</td>
      <TD ALIGN="CENTER" HEIGHT="20"> 85</TD>
      <TD ALIGN="CENTER"> 85</TD>
      <TD ALIGN="CENTER"> 85</TD>
      <TD ALIGN="CENTER"> 80</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR> 
      <td bgcolor="#cccccc">10</td>
      <TD ALIGN="CENTER" HEIGHT="20"> 65</TD>
      <TD ALIGN="CENTER"> 60</TD>
      <TD ALIGN="CENTER"> 65</TD>
      <TD ALIGN="CENTER"> 60</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR> 
      <td bgcolor="#cccccc">11</td>
      <TD ALIGN="CENTER" HEIGHT="20"> 75</TD>
      <TD ALIGN="CENTER"> 65</TD>
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER"> 70</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR> 
      <td bgcolor="#cccccc">12</td>
      <TD ALIGN="CENTER" HEIGHT="20"> 85</TD>
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER"> 70</TD>
      <TD ALIGN="CENTER"> 75</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR> 
      <td bgcolor="#cccccc">13</td>
      <TD ALIGN="CENTER" HEIGHT="20"> 80</TD>
      <TD ALIGN="CENTER"> 70</TD>
      <TD ALIGN="CENTER"> 65</TD>
      <TD ALIGN="CENTER"> 70</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR> 
      <td bgcolor="#cccccc">14</td>
      <TD ALIGN="CENTER" HEIGHT="20"> 85</TD>
      <TD ALIGN="CENTER"> 80</TD>
      <TD ALIGN="CENTER"> 80</TD>
      <TD ALIGN="CENTER"> 80</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <TR> 
      <td bgcolor="#cccccc">15</td>
      <TD ALIGN="CENTER" HEIGHT="20"> 90</TD>
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER"> 85</TD>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
      <td bgcolor="#cccccc"></td>
    </TR>
    <tr><td bgcolor="#cccccc">平均</td><td>( a )</td><td>( b )</td><td>( c )</td><td>( d )</td>
        <td>( e )</td><td>( f )</td><td>( g )</td><td> ( h )</td><td>( i )</td>
    </tr>
    <tr><td bgcolor="#cccccc">偏差平方和</td><td>( j )</td><td>( k )</td><td>( l )</td><td>( m )</td>
        <td bgcolor="#cccccc"></td><td bgcolor="#cccccc"></td><td bgcolor="#cccccc"></td><td bgcolor="#cccccc"></td><td>( n )</td>
    </tr>
    </TABLE>

In [6]:
# 以下の４つをそれぞれ「群」と呼びます。
a1b1 = [65, 85, 75, 85, 75, 80, 90, 75, 85, 65, 75, 85, 80, 85, 90] # クリスピーで辛口
a1b2 = [65, 70, 80, 75, 70, 60, 65, 70, 85, 60, 65, 75, 70, 80, 75] # クリスピーで普通味
a2b1 = [70, 65, 85, 80, 75, 65, 75, 60, 85, 65, 75, 70, 65, 80, 75] # 普通の衣で辛口
a2b2 = [70, 70, 85, 80, 65, 75, 65, 85, 80, 60, 70, 75, 70, 80, 85] # 普通の衣で普通味

In [7]:
# 以下の４つのうち、上２つを「要因A（食感）」、下２つを「要因B（味）」と呼びます。
a1 = a1b1 + a1b2 # クリスピーのデータを作ってください。
a2 = a2b1 + a2b2 # 普通の衣のデータを作ってください。
b1 = a1b1 + a2b1 # 辛口のデータを作ってください。
b2 = a1b2 + a2b2 # 普通味のデータを作ってください。

* 各項目の平均値を計算してください（上表 a, b, c, d, e, f, g, h, i ）

In [8]:
# 各項目の平均値を計算してください（上表 a, b, c, d, e, f, g, h, i ）
zentai = a1b1 + a1b2 + a2b1 + a2b2
a = average(a1b1)
b = average(a1b2)
c = average(a2b1)
d = average(a2b2)
e = average(a1)
f = average(a2)
g = average(b1)
h = average(b2)
i = average(zentai)
print (a, b, c, d)
print (e, f, g, h)
print (i)

79.66666666666667 71.0 72.66666666666667 74.33333333333333
75.33333333333333 73.5 76.16666666666667 72.66666666666667
74.41666666666667


* 各群の偏差平方和を計算してください（上表 j, k, l, m ）。

In [9]:
def sum_square_deviation(list):
    ave = average(list)
    sum = 0
    for x in list:
        sum += (x - ave)**2
    return sum

In [10]:
# 各群の偏差平方和を計算してください（上表 j, k, l, m ）
j = sum_square_deviation(a1b1)
k = sum_square_deviation(a1b2)
l = sum_square_deviation(a2b1)
m = sum_square_deviation(a2b2)
print (j, k, l, m)

873.3333333333333 760.0 843.3333333333333 893.3333333333333


* 全体の偏差平方和を計算してください（上表 n ）。

In [11]:
# 全体の偏差平方和を計算してください（上表 n ）
n = sum_square_deviation(zentai)
print (n)

4004.5833333333326


__要因Aによる残差（偏差平方和）を計算してください__ （下表 o )

* 要因A（食感）による残差 ＝ (クリスピーの平均 ー 全体の平均)の２乗 ✕ クリスピーのデータ個数　＋　(普通の衣の平均 ー 全体の平均)の２乗　✕ 普通衣のデータ個数

In [12]:
# 要因Aによる残差（偏差平方和）を計算してください （下表 o )
o = (average(a1) - average(zentai))**2 * len(a1) + (average(a2) - average(zentai))**2 * len(a2)
print(o)

50.4166666666664


__要因Bによる残差（偏差平方和）を計算してください__ (下表 p )

* 要因B（味）による残差 ＝ (辛口の平均 ー 全体の平均)の２乗 ✕ 辛口のデータ個数　＋　(普通味の平均 ー 全体の平均)の２乗　✕ 普通味のデータ個数

In [13]:
# 要因Bによる残差（偏差平方和）を計算してください (下表 p )
p = (average(b1) - average(zentai))**2 * len(b1) + (average(b2) - average(zentai))**2 * len(b2)
print(p)

183.75


__交互作用による残差（偏差平方和）を計算してください__ (下表 q )

* 各群の平均の残差＝ Σ(（各群の平均　ー　全体の平均）の2乗✕ 各群のデータ個数)
* 交互作用による残差　=　各群の平均の残差　－　要因Aによる残差　－　要因Bによる残差

In [14]:
# 交互作用による残差（偏差平方和）を計算してください (下表 q )
q = (average(a1b1) - average(zentai))**2 * len(a1b1) \
    + (average(a1b2) - average(zentai))**2 * len(a1b2) \
    + (average(a2b1) - average(zentai))**2 * len(a2b1) \
    + (average(a2b2) - average(zentai))**2 * len(a2b2) \
    - o - p
print(q)

400.41666666666754


__残りの残差（偏差平方和）を計算してください__ (下表 r )

* 残りの残差とは、群内の偏差平方和を足したものです（ j + k + l + m ）。

In [15]:
# 残りの残差（偏差平方和）を計算してください (下表 r )
r = j + k + l + m
print(r)

3370.0


__分散分析表を作る__

o, p, q, r, n は既に計算済みです。

<TABLE BORDER="1" align="center" CELLPADDING="2" CELLSPACING="0">
  <TR> 
    <TD WIDTH=100 BGCOLOR="#CCCCCC"> <div align="center">要因 </div></TD>
    <TD WIDTH=100 BGCOLOR="#DDDDDD"> <CENTER>
       偏差平方和 </CENTER></TD>
    <TD WIDTH=100 BGCOLOR="#DDDDDD"> <CENTER>
        自由度 </CENTER></TD>
    <TD WIDTH=100 BGCOLOR="#DDDDDD"> <CENTER>
        平均平方 </CENTER></TD>
    <TD WIDTH=100 BGCOLOR="#DDDDDD"> <CENTER>
        F </CENTER></TD>
  </TR>
  <TR> 
    <TD WIDTH=100 BGCOLOR="#DDDDDD"> <div align="center">要因A </div></TD>
    <TD WIDTH=100> <div align="right">( o )</div></TD>
    <TD WIDTH=100> <div align="right">( s )</div></TD>
    <TD WIDTH=100> <div align="right">( x )</div></TD>
    <TD WIDTH=100> <div align="right">( bb )</div></TD>
  </TR>
  <TR> 
    <TD WIDTH=100 BGCOLOR="#DDDDDD"> <div align="center">要因B </div></TD>
    <TD WIDTH=100> <div align="right">( p )</div></TD>
    <TD WIDTH=100> <div align="right">( t )</div></TD>
    <TD WIDTH=100> <div align="right">( y )</div></TD>
    <TD WIDTH=100> <div align="right">( cc )</div></TD>
  </TR>
  <TR> 
    <TD WIDTH=100 BGCOLOR="#dddddd"> <div align="center">交互作用 </div></TD>
    <TD WIDTH=100> <div align="right">( q )</div></TD>
    <TD WIDTH=100> <div align="right">( u )</div></TD>
    <TD WIDTH=100> <div align="right">( z )</div></TD>
    <TD WIDTH=100> <div align="right">( dd )</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#dddddd"> <div align="center">残差</div></TD>
    <TD><div align="right">( r )</div></TD>
    <TD><div align="right">( w )</div></TD>
    <TD><div align="right">( aa )</div></TD>
  </TR>
  <TR> 
    <TD BGCOLOR="#CCCCCC"><div align="center">全体 </div></TD>
    <TD><div align="right">( n )</div></TD>
    <TD><div align="right">( v )</div></TD>
  </TR>
</TABLE>

* 自由度を求めてください。
  * (上表 s ) 要因Aの自由度 ＝ 要因Aの中の条件数（群の数）から１を引いたもの
  * (上表 t ) 要因Bの自由度 ＝ 要因Bの中の条件数（群の数）から１を引いたもの
  * (上表 u ) 交互作用の自由度　＝ 要因Aの自由度 ✕ 要因Bの自由度
  * (上表 v ) 全体の自由度　＝ 全データ数から１を引いたもの
  * (上表 w ) 残差の自由度 ＝ 全体の自由度 ー 要因Aの自由度 ー 要因Bの自由度 ー 交互作用の自由度

In [16]:
# 自由度を求めてください。
s = 2 - 1
t = 2 - 1
u = s * t
v = len(zentai) - 1
w = v - s - t - u
print (s, t, u, w, v)

1 1 1 56 59


* 要因A、要因B、交互作用、残差の平均平方をそれぞれ求めてください。(上表 x, y, z, aa )
  * 平均平方 ＝ 偏差平方和 ／ 自由度

In [17]:
# 要因A、要因B、交互作用、残差の平均平方をそれぞれ求めてください。(上表 x, y, z, aa )
x = o / s
y = p / t
z = q / u
aa = r / w
print (x, y, z, aa)

50.4166666666664 183.75 400.41666666666754 60.17857142857143


* F値を求めてください。
  * (上表 bb ) 要因AのF値 ＝ 要因Aの平均平方 ／ 残差の平均平方
  * (上表 cc ) 要因BのF値 ＝ 要因Bの平均平方 ／ 残差の平均平方
  * (上表 dd ) 交互作用のF値 ＝ 交互作用の平均平方 ／ 残差の平均平方

In [18]:
# F値を求めてください。
bb = x / aa
cc = y / aa
dd = z / aa
print (bb, cc, dd)

0.8377843719089966 3.0534124629080117 6.653808110781418


* 計算したF値と、F分布表を比較して、以下の問いに答えてください。
  * 要因A（食感）による点数の差があるかどうか答えてください。
  * 要因B（味）　による点数の差があるかどうか答えてください。
  * 交互作用による点数の差があるかどうか答えてください。

In [19]:
# 計算したF値と、F分布表を比較

# 要因1（食感）による点数の差があるかどうか答えてください。

# 要因1では、群内の自由度が56（近いところで60）、群間の自由度が1になります。
# F分布表を見ると、5%有意水準で、F=4.00、1%有意水準で、F=7.08、となりますので、
# 要因1のF=0.84　というのは、いずれの棄却域にも入りません。
# したがって、「食感の要因による点数の差はない」と結論できます。

# 要因2（味）　による点数の差があるかどうか答えてください。

# 要因２でも、群内の自由度が56（近いところで60）、群間の自由度が1になります。
# F分布表を見ると、5%有意水準で、F=4.00、1%有意水準で、F=7.08、となりますので、
# 要因2のF=3.05　というのは、いずれの棄却域にも入りません。
# したがって、「味の要因による点数の差はない」と結論できます。

# 交互作用による点数の差があるかどうか答えてください。

# 交互作用でも、群内の自由度が56（近いところで60）、群間の自由度が1になります。
# F分布表を見ると、5%有意水準で、F=4.00、1%有意水準で、F=7.08、となりますので、
# 交互作用のF=6.65　というのは、5%有意水準において棄却域に入ります。
# 以上から、要因１要因２の単独の効果はないが、組み合わせによる効果はある（点数の差がある）と言えます。

<h2 STYLE="background: #c2edff;padding: 0.5em;"><a href="http://kogolab.chillout.jp/elearn/hamburger/chap7/sec3.html" target="_blank">7.3　交互作用とは?</a></h2>

<h4 style="border-bottom: solid 1px black;">交互作用が有意</h4>

分散分析の結果、２つの要因による主効果には有意差が見られず、交互作用に有意差が見られました。

主効果に有意差がないということは、「その要因単独の効果はない」ということです。

各群の平均を見ると、

<TABLE WIDTH="342" HEIGHT="71" BORDER="1" align="center" CELLPADDING=
  "0" CELLSPACING="0">
  <TR> 
    <TD WIDTH="30%" HEIGHT="20" ALIGN="CENTER">　 </TD>
    <TD WIDTH="35%" ALIGN="CENTER" BGCOLOR="#eeee77">クリスピー</TD>
    <TD WIDTH="35%" ALIGN="CENTER" BGCOLOR="#bbbbbb">普通の衣</TD>
  </TR>
  <TR> 
    <TD HEIGHT="20" ALIGN="CENTER" BGCOLOR="#ff8888">辛口</TD>
    <TD ALIGN="CENTER"> 　79.67</TD>
    <TD ALIGN="CENTER"> 　72.67</TD>
  </TR>
  <TR> 
    <TD HEIGHT="20" ALIGN="CENTER" BGCOLOR="#dddddd">普通味</TD>
    <TD ALIGN="CENTER"> 　71.00</TD>
    <TD ALIGN="CENTER"> 　74.33</TD>
  </TR>
</TABLE>

辛口に関していえば……クリスピーと普通の衣を比べると、クリスピーの方が高い。

<table><tr><td><img src="image/chicken00002.jpg"></td>
<td><</td>
<td><img src="image/chicken00003.jpg"></td>
</tr></table>

普通味に関していえば……クリスピーと普通の衣を比べると、普通の方が高い。

<table><tr><td><img src="image/chicken00004.jpg"></td>
<td>></td>
<td><img src="image/chicken00001.jpg"></td>
</tr></table>

ということは、「クリスピー、辛口は、単独で使うより、両方を使った方がよく、逆に、どちらかを使うくらいなら、普通のチキンの方がよい」ということが分かります。

上の結果を、あなたは店員さんに伝えました。

「そうか。それぞれを単独で売るくらいならやめて、両方を組み合わせて売ったほうがいいんだ。よし、辛口クリスピーチキンを売り出すことにするよ！」

<img src="image/salesman00020.jpg">

<h4 style="border-bottom: solid 1px black;">交互作用の意味</h4>

  <P>交互作用の意味について、図式的に説明します。</P>
  
<P>交互作用のグラフは、縦軸に点数を、横軸に条件をとり、群別に平均点をとります。</P>
<P>下のグラフで、点数は表示されていませんが、a1、a2が要因Aによる群、b1、b2が要因Bによる群、<FONT COLOR="#FF0000">●</FONT>や<FONT COLOR="#99CCFF">■</FONT>で表示されているのが平均点として読みとってください。</P>

<h4 style="border-bottom: solid 1px black;">交互作用のない場合</h4>

まず、交互作用のない場合を見てみましょう。

交互作用のない場合のグラフは、グラフは平行になります。

<img src="image/fig_kougo_nashi.gif">

<h4 style="border-bottom: solid 1px black;">交互作用のある場合</h4>

交互作用のある場合のグラフは平行にはなりません。

以下に交互作用のある場合のグラフを示します。

<img src="image/fig_kougo_ari.gif">

<P>ここでは、各群の平均が</P>
<TABLE WIDTH="342" HEIGHT="71" BORDER="1" align="center" CELLPADDING=
  "0" CELLSPACING="0">
  <TR> 
    <TD WIDTH="30%" HEIGHT="20" ALIGN="CENTER">　 </TD>
    <TD WIDTH="35%" ALIGN="CENTER" BGCOLOR="#eeee77">クリスピー</TD>
    <TD WIDTH="35%" ALIGN="CENTER" BGCOLOR="#bbbbbb">普通の衣</TD>
  </TR>
  <TR> 
    <TD HEIGHT="20" ALIGN="CENTER" BGCOLOR="#ff8888">辛口</TD>
    <TD ALIGN="CENTER"> 　79.67</TD>
    <TD ALIGN="CENTER"> 　72.67</TD>
  </TR>
  <TR> 
    <TD HEIGHT="20" ALIGN="CENTER" BGCOLOR="#dddddd">普通味</TD>
    <TD ALIGN="CENTER"> 　71.00</TD>
    <TD ALIGN="CENTER"> 　74.33</TD>
  </TR>
</TABLE>
<P>
<P>このようになっていました。これをグラフで表してみると、</P>
 
<img src="image/fig_kougo_x.gif">

これは(6)のケースにあたり、要因A（クリスピーか普通の衣か）の主効果も、要因B（辛口か普通味か）の主効果もないけれども、クリスピーのときは辛口がよく、普通の衣のときは普通味のほうがよいという交互作用がでているのです。

つまり、組み合わされる要因によって効果の現れ方が違うということです。これが交互作用です。

<h2 STYLE="background: #c2edff;padding: 0.5em;"><a href="http://kogolab.chillout.jp/elearn/hamburger/chap7/sec4.html" target="_blank">7.4　実践編：セットでいこう</a></h2>

<h4 style="border-bottom: solid 1px black;">ピリ辛セット</h4>

辛口クリスピーチキンは大好評で、あなたの分析が功を奏したようです。

「辛口、ってのがうけてるみたいなんだよ、それに、モグモグもパクパクもチキンには力を入れてないみたいだからね」

<img src="image/salesman00027.jpg">

あなたが、ちらっと厨房のなかを覗くと、店員さんたちがハンバーガーとポテトを試食していました。

「今、食事なんですか？」

「いや、あれはね、新しいセット商品の試食してるんだ。辛いのがうけたから、ピリ辛バーガーとピリ辛ポテトのセットを売り出そうっていう案が出てね。」

どうやら、ワクワクでは「ピリ辛セット」を売り出したいみたいです。

「ピリ辛セットかぁ・・・でも『辛い×辛い』でいいんですか？辛いのがうけたからといって、単純にそうするのは安易のような気がするけど・・・」

あなたは、セットの組み合わせを検討するために、チキンのときと同じように、街の人60人に、それぞれの組み合わせセットに100点満点で点数をつけてもらいました。

<h4 style="padding: 0.25em 0.5em;color: #494949;background: transparent;border-left: solid 5px #7db4e6;">課題7.4</h4>

次のデータで同様に分散分析をしてください。
<TABLE WIDTH="403" HEIGHT="351" BORDER="1" CELLSPACING="0" 
  CELLPADDING="0">
    <TR> 
      <TD height="29" COLSPAN="2" BGCOLOR="#ff9c05"> 
        <CENTER>
          ピリ辛バーガー </CENTER></TD>
      <TD COLSPAN="2" BGCOLOR="#f97c6a"> <CENTER>
          普通のハンバーガー </CENTER></TD>
    </TR>
    <TR> 
      <TD WIDTH="20%" BGCOLOR="#fb3a04" HEIGHT="20"> <CENTER>
          ピリ辛ポテト </CENTER></TD>
      <TD WIDTH="20%" BGCOLOR="#ffd303"> <CENTER>
          普通のポテト </CENTER></TD>
      <TD WIDTH="20%" BGCOLOR="#fa3a04"> <CENTER>
          ピリ辛ポテト </CENTER></TD>
      <TD WIDTH="20%" BGCOLOR="#ffd303"> <CENTER>
          普通のポテト </CENTER></TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER"> 80</TD>
      <TD ALIGN="CENTER"> 85</TD>
      <TD ALIGN="CENTER" HEIGHT="20"> 60</TD>
      <TD ALIGN="CENTER"> 55</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER"> 85</TD>
      <TD ALIGN="CENTER"> 65</TD>
      <TD ALIGN="CENTER" HEIGHT="20"> 70</TD>
      <TD ALIGN="CENTER"> 65</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER"> 80</TD>
      <TD ALIGN="CENTER" HEIGHT="20"> 80</TD>
      <TD ALIGN="CENTER"> 50</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER"> 70</TD>
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER" HEIGHT="20"> 75</TD>
      <TD ALIGN="CENTER"> 70</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER"> 85</TD>
      <TD ALIGN="CENTER" HEIGHT="20"> 65</TD>
      <TD ALIGN="CENTER"> 50</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER"> 60</TD>
      <TD ALIGN="CENTER"> 65</TD>
      <TD ALIGN="CENTER" HEIGHT="20"> 60</TD>
      <TD ALIGN="CENTER"> 65</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER"> 70</TD>
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER" HEIGHT="20"> 70</TD>
      <TD ALIGN="CENTER"> 75</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER"> 60</TD>
      <TD ALIGN="CENTER" HEIGHT="20"> 75</TD>
      <TD ALIGN="CENTER"> 65</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER"> 85</TD>
      <TD ALIGN="CENTER"> 80</TD>
      <TD ALIGN="CENTER" HEIGHT="20"> 80</TD>
      <TD ALIGN="CENTER"> 75</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER"> 70</TD>
      <TD ALIGN="CENTER"> 70</TD>
      <TD ALIGN="CENTER" HEIGHT="20"> 65</TD>
      <TD ALIGN="CENTER"> 65</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER" HEIGHT="20"> 65</TD>
      <TD ALIGN="CENTER"> 70</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER"> 85</TD>
      <TD ALIGN="CENTER"> 80</TD>
      <TD ALIGN="CENTER" HEIGHT="20"> 75</TD>
      <TD ALIGN="CENTER"> 55</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER"> 65</TD>
      <TD ALIGN="CENTER" HEIGHT="20"> 70</TD>
      <TD ALIGN="CENTER"> 65</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER"> 85</TD>
      <TD ALIGN="CENTER"> 85</TD>
      <TD ALIGN="CENTER" HEIGHT="20"> 85</TD>
      <TD ALIGN="CENTER"> 70</TD>
    </TR>
    <TR> 
      <TD ALIGN="CENTER"> 90</TD>
      <TD ALIGN="CENTER"> 75</TD>
      <TD ALIGN="CENTER" HEIGHT="20"> 60</TD>
      <TD ALIGN="CENTER"> 65</TD>
    </TR>
  </TABLE>

In [20]:
# 以下の４つをそれぞれ「群」と呼びます。
a1b1 = [80, 85, 75, 70, 75, 60, 70, 75, 85, 70, 75, 85, 75, 85, 90] # ピリ辛バーガー + ピリ辛ポテト
a1b2 = [85, 65, 80, 75, 85, 65, 75, 60, 80, 70, 75, 80, 65, 85, 75] # ピリ辛バーガー + 普通のポテト
a2b1 = [60, 70, 80, 75, 65, 60, 70, 75, 80, 65, 65, 75, 70, 85, 60] # 普通のハンバーガー + ピリ辛ポテト
a2b2 = [55, 65, 50, 70, 50, 65, 75, 65, 75, 65, 70, 55, 65, 70, 65] # 普通のハンバーガー + 普通のポテト

In [21]:
# 以下の４つのうち、上２つを「要因１（バーガーの味）」、下２つを「要因２（ポテトの味）」と呼びます。
a1 = a1b1 + a1b2 # ピリ辛バーガーのデータを作ってください。
a2 = a2b1 + a2b2 # 普通のバーガーのデータを作ってください。
b1 = a1b1 + a2b1 # ピリ辛ポテトのデータを作ってください。
b2 = a1b2 + a2b2 # 普通のポテトのデータを作ってください。

In [22]:
# 各項目の平均値を計算してください（上表 a, b, c, d, e, f, g, h, i ）
zentai = a1b1 + a1b2 + a2b1 + a2b2
a = average(a1b1)
b = average(a1b2)
c = average(a2b1)
d = average(a2b2)
e = average(a1)
f = average(a2)
g = average(b1)
h = average(b2)
i = average(zentai)
print (a, b, c, d)
print (e, f, g, h)
print (i)

77.0 74.66666666666667 70.33333333333333 64.0
75.83333333333333 67.16666666666667 73.66666666666667 69.33333333333333
71.5


In [23]:
# 各群の偏差平方和を計算してください（上表 j, k, l, m ）
j = sum_square_deviation(a1b1)
k = sum_square_deviation(a1b2)
l = sum_square_deviation(a2b1)
m = sum_square_deviation(a2b2)
print (j, k, l, m)

890.0 923.3333333333334 873.3333333333334 910.0


In [24]:
# 全体の偏差平方和を計算してください（上表 n ）
n = sum_square_deviation(zentai)
print (n)

5065.0


In [25]:
# 要因Aによる残差（偏差平方和）を計算してください （下表 o )
o = (average(a1) - average(zentai))**2 * len(a1) + (average(a2) - average(zentai))**2 * len(a2)
print(o)

1126.6666666666642


In [26]:
# 要因Bによる残差（偏差平方和）を計算してください (下表 p )
p = (average(b1) - average(zentai))**2 * len(b1) + (average(b2) - average(zentai))**2 * len(b2)
print(p)

281.6666666666679


In [27]:
# 交互作用による残差（偏差平方和）を計算してください (下表 q )
q = (average(a1b1) - average(zentai))**2 * len(a1b1) \
    + (average(a1b2) - average(zentai))**2 * len(a1b2) \
    + (average(a2b1) - average(zentai))**2 * len(a2b1) \
    + (average(a2b2) - average(zentai))**2 * len(a2b2) \
    - o - p
print(q)

60.00000000000182


In [28]:
# 残りの残差（偏差平方和）を計算してください (下表 r )
r = j + k + l + m
print(r)

3596.666666666667


In [29]:
# 自由度を求めてください。
s = 2 - 1
t = 2 - 1
u = s * t
v = len(zentai) - 1
w = v - s - t - u
print (s, t, u, w, v)

1 1 1 56 59


In [30]:
# 要因A、要因B、交互作用、残差の平均平方をそれぞれ求めてください。(上表 x, y, z, aa )
x = o / s
y = p / t
z = q / u
aa = r / w
print (x, y, z, aa)

1126.6666666666642 281.6666666666679 60.00000000000182 64.22619047619048


In [31]:
# F値を求めてください
bb = x / aa
cc = y / aa
dd = z / aa
print (bb, cc, dd)

17.542168674698758 4.385542168674717 0.9341983317887215


In [32]:
# 計算したF値と、F分布表を比較

# 要因1：自由度は1と56、F=17.54。これは1%水準で有意
# 要因2：自由度は1と56、F=4.39。これは5%水準で有意
# 交互作用：自由度は1と56、F=0.93。これは有意ではない　
# この分散分析の結果と、平均値のグラフを読みとると、
# ポテトもハンバーガーもセットの組み合わせによる影響がなく、
# それぞれの「ピリ辛」が高い評価を得たということがわかります。
#したがって、ピリ辛ポテトにピリ辛バーガーの組み合わせが最も高い評価を得ました。

分散分析の結果、要因A、要因Bの主効果に有意差が見られ、交互作用には有意差が見られませんでした。

この分散分析の結果と、平均値のグラフを読みとると、ポテトもハンバーガーもセットの組み合わせによる影響がなく、それぞれの「ピリ辛」が高い評価を得たということがわかります。したがって、ピリ辛ポテトにピリ辛バーガーの組み合わせが最も高い評価を得ました。

<img src="image/salesman00026.jpg">

「よし、ピリ辛セットでいくぞ！」

<h2 STYLE="background: #c2edff;padding: 0.5em;">ハンバーガー統計学 第７章 課題</h2>

__課題7.4__ を解いて、指定のメールアドレスまでメールしてください。メール送信後は、エラーが帰ってきてないことを確認してください（メールアドレスを間違える人がときどき居ます）。
* 締切：10月17日（今すぐでなくても結構です）
* メールタイトル：「ハンバーガー統計学 第７章」
* 学籍番号と氏名を明記すること。
* 感想などがあれば書いてくれると嬉しいです。次回以降の講義の改善につながるかも知れません。

In [33]:
# ハンバーガー統計学 第７章 課題

<h2 STYLE="background: #c2edff;padding: 0.5em;">終わったら、<a href="HamburgerStatistics08.ipynb">第８章：分布</a> に進んでください。</h2>
（2017年10月17日講義、10月20日課題提出締切）